### ISSUE TITLE
L014 linting and fixing "Inconsistent capitalisation of unquoted identifiers" that do not exist

### ISSUE THREAD

**Issue by GClunies** [sqlfluff/sqlfluff#966]
## Expected Behaviour
If the SQL has _consistent_ unquoted identifiers that matches the policy in the `.sqlfluff` config, for example:
```
[sqlfluff:rules:L014]  # Unquoted identifiers
capitalisation_policy = lower
unquoted_identifiers_policy = all
```
Then  `sqlfluff` should not throw any `L014 | Inconsistent capitalisation of unquoted identifiers` errors and should not apply any `L014` "fixes" (since they do not exist). 

## Observed Behaviour
1. `sqlfluff lint sl_web_spot_pages.sql` is finding `L014 | Inconsistent capitalisation of unquoted identifiers` linting errors that do not exist.
2. `sqlfluff fix sl_web_spot_pages.sql` is "fixing" `L014 | Inconsistent capitalisation of unquoted identifiers` errors that do not exist. These "fixes" are not desired and result in inconsistent casing throughout the "fixed" SQL. I want to maintain `lower` casing throughout my dbt models.

## Version
I have installed sqlfluff from source using `pip install git+https://github.com/sqlfluff/sqlfluff.git` (currently on commit: 9970bf2e7001b23dc2f84c5ec572a28cd92099f6) so that I can take advantage of recent PRs (#868 and #964) instead of having to wait for the next release.

<details>
<summary>Source dbt model SQL</summary>
<p>

```sql
-- filename: sl_web_spot_pages.sql

{{
	config(
        materialized = 'incremental',
        sort = 'received_at',
        unique_key = 'page_id'
	)
}}

with user_entitlement as (

    select * from {{ ref('sl_web_user_entitlement_base') }}

),

has_cam_entitlement as (

    select * from {{ ref('spots_has_cam_map') }}

),

pages as (

    select * from {{ ref('base_web_pages') }}

),

pages_xf as (

    select
        pages.received_at,
        pages.page_id,
        pages.spot_id,
        pages.user_id,
        pages.anonymous_id,
        pages.cam_id,
        pages.is_single_cam,
        pages.is_mobile_view,
        pages.cam_name_primary,
        user_entitlement.subscription_entitlement

    from pages
    left join user_entitlement
        on pages.user_id = user_entitlement.user_id
            and pages.received_at >= user_entitlement.entitlement_from
            and pages.received_at < user_entitlement.entitlement_to
    where pages.spot_id is not null
        and pages.name = 'Spot Report'

        -- Remove multicam favorites page hits for now due to varying # of cams on page (user selected)
        -- and pages.context_page_url != 'https://www.surfline.com/surf-cams'
        /*
        2019-12-28
        There are only 13 records of the above "&" scenario - which is immaterial.
        This might not be worth the compute time.

        - Julian
        */

    {% if is_incremental() %}

        -- Only get new data (not including today's data)
        and pages.received_at > (select max (received_at) from {{ this }} )
        and pages.received_at < current_date

    {% else %}

        and pages.received_at >= date_trunc('year', current_date - 1) - interval '1 year'

    {% endif %}

),

final as (

    select
        pages_xf.received_at,
        pages_xf.page_id,
        pages_xf.spot_id,
        has_cam_entitlement.has_cam,
        pages_xf.user_id,
        pages_xf.anonymous_id,
        pages_xf.cam_id,
        pages_xf.is_single_cam,
        pages_xf.is_mobile_view,
        pages_xf.cam_name_primary,
        pages_xf.subscription_entitlement

    from pages_xf
    -- inner join because some spot_ids not in MongoDB
    inner join has_cam_entitlement using (spot_id)

)

select * from final

```
</p>
</details>

<details>
<summary>Compiled dbt model SQL (from source above using dbt-compiler)</summary>
<p>

```sql
-- filename: sl_web_spot_pages.sql



with  __dbt__CTE__sl_web_user_entitlement_base as (
-- filename: sl_web_user_entitlement_base.sql



/*
Rows in this model are:
    - Individual hits of Segment identify() calls on web or native.

This model can be used to map user entitlement to pages or segment events. This
model does NOT summarize the "subscription window(s)" for a given user_id and 
subscription. Further modeling and transformations are required to extract 
subscription window summaries.
*/

with final as (

    select distinct  -- Make sure duplicate rows are not included
        user_id,
        subscription_entitlement,
        received_at as entitlement_from,
        coalesce(lead(received_at, 1) over(partition by user_id order by received_at),'3001-01-01') as entitlement_to
                
    from surfline.identifies

    where subscription_entitlement is not null

)

select * from final
),  __dbt__CTE__base_web_pages as (
--filename: base_web_pages.sql



with final as (

    select
        received_at,
        id as page_id,
        spot_id,
        has_cam,
        user_id,
        anonymous_id,
        name,
        -- fields below are in production as of: 2020-01-31 (1st full day of data)
        cam_id,
        is_single_cam,
        is_mobile_view,
        cam_name_primary

    from surfline.pages

)

select * from final
),user_entitlement as (

    select * from __dbt__CTE__sl_web_user_entitlement_base

),

has_cam_entitlement as (

    select * from "segment"."greg_clunies"."spots_has_cam_map"

),

pages as (

    select * from __dbt__CTE__base_web_pages

),

pages_xf as (

    select
        pages.received_at,
        pages.page_id,
        pages.spot_id,
        pages.user_id,
        pages.anonymous_id,
        pages.cam_id,
        pages.is_single_cam,
        pages.is_mobile_view,
        pages.cam_name_primary,
        user_entitlement.subscription_entitlement

    from pages
    left join user_entitlement
        on pages.user_id = user_entitlement.user_id
            and pages.received_at >= user_entitlement.entitlement_from
            and pages.received_at < user_entitlement.entitlement_to
    where pages.spot_id is not null
        and pages.name = 'Spot Report'

        -- Remove multicam favorites page hits for now due to varying # of cams on page (user selected)
        -- and pages.context_page_url != 'https://www.surfline.com/surf-cams'
        /*
        2019-12-28
        There are only 13 records of the above "&" scenario - which is immaterial.
        This might not be worth the compute time.

        - Julian
        */

    

        -- Only get new data (not including today's data)
        and pages.received_at > (select max (received_at) from "segment"."greg_clunies"."sl_web_spot_pages" )
        and pages.received_at < current_date

    

),

final as (

    select
        pages_xf.received_at,
        pages_xf.page_id,
        pages_xf.spot_id,
        has_cam_entitlement.has_cam,
        pages_xf.user_id,
        pages_xf.anonymous_id,
        pages_xf.cam_id,
        pages_xf.is_single_cam,
        pages_xf.is_mobile_view,
        pages_xf.cam_name_primary,
        pages_xf.subscription_entitlement

    from pages_xf
    -- inner join because some spot_ids not in MongoDB
    inner join has_cam_entitlement using (spot_id)

)

select * from final
```
</p>
</details>

<details>
<summary>"Fixed" dbt model SQL (results in inconsistent casing)</summary>
<p>

```sql
-- filename: sl_web_spot_pages.sql

{{
	config(
        materialized = 'incremental',
        sort = 'received_at',
        unique_key = 'page_id'
	)
}}

with user_entitlement as (

    select * from {{ ref('sl_web_user_entitlement_base') }}

),

HAS_CAM_ENTITLEMENT as (

    select * from {{ ref('spots_has_cam_map') }}

),

PAGES as (

    select * from {{ ref('base_web_pages') }}

),

PAGES_XF as (

    select
        PAGES.RECEIVED_AT,
        PAGES.PAGE_ID,
        PAGES.SPOT_ID,
        PAGES.USER_ID,
        PAGES.ANONYMOUS_ID,
        PAGES.CAM_ID,
        PAGES.IS_SINGLE_CAM,
        PAGES.IS_MOBILE_VIEW,
        PAGES.CAM_NAME_PRIMARY,
        USER_ENTITLEMENT.SUBSCRIPTION_ENTITLEMENT

    from PAGES
    left join USER_ENTITLEMENT
        on PAGES.USER_ID = USER_ENTITLEMENT.USER_ID
            and PAGES.RECEIVED_AT >= USER_ENTITLEMENT.ENTITLEMENT_FROM
            and PAGES.RECEIVED_AT < USER_ENTITLEMENT.ENTITLEMENT_TO
    where PAGES.SPOT_ID is not null
        and PAGES.NAME = 'Spot Report'

        -- Remove multicam favorites page hits for now due to varying # of cams on page (user selected)
        -- and pages.context_page_url != 'https://www.surfline.com/surf-cams'
        /*
        2019-12-28
        There are only 13 records of the above "&" scenario - which is immaterial.
        This might not be worth the compute time.

        - Julian
        */

    {% if is_incremental() %}

        -- Only get new data (not including today's data)
        and PAGES.RECEIVED_AT > (select max(RECEIVED_AT) from {{ this }} )
        and PAGES.RECEIVED_AT < current_date

    {% else %}

        and pages.received_at >= date_trunc('year', current_date - 1) - interval '1 year'

    {% endif %}

),

FINAL as (

    select
        PAGES_XF.RECEIVED_AT,
        PAGES_XF.PAGE_ID,
        PAGES_XF.SPOT_ID,
        HAS_CAM_ENTITLEMENT.HAS_CAM,
        PAGES_XF.USER_ID,
        PAGES_XF.ANONYMOUS_ID,
        PAGES_XF.CAM_ID,
        PAGES_XF.IS_SINGLE_CAM,
        PAGES_XF.IS_MOBILE_VIEW,
        PAGES_XF.CAM_NAME_PRIMARY,
        PAGES_XF.SUBSCRIPTION_ENTITLEMENT

    from PAGES_XF
    -- inner join because some spot_ids not in MongoDB
    inner join HAS_CAM_ENTITLEMENT using (SPOT_ID)

)

select * from FINAL

```
</p>
</details>

## Configuration
<details>
<summary>SQLFluff configuration</summary>
<p>

```
# For SQLFluff Rules reference, see:
# https://docs.sqlfluff.com/en/stable/rules.html#rules-reference
[sqlfluff]
verbose = 0
nocolor = False
dialect = postgres
templater = dbt
rules = None
exclude_rules = L032,L033,L034,L037,L044
recurse = 0
output_line_length = 120
runaway_limit = 10
ignore = parsing
ignore_templated_areas = True

[sqlfluff:indentation]
indented_joins = False
template_blocks_indent = True

# Some rules can be configured directly from the config common to other rules.
[sqlfluff:rules]
tab_space_size = 4
max_line_length = 120
indent_unit = space
comma_style = trailing
allow_scalar = True
single_table_references = consistent
only_aliases = True

# Some rules have their own specific config.
# All SQLFluff rules can be found at: https://docs.sqlfluff.com/en/stable/rules.html#rules-reference
# When a rule in not listed below, we inherit the default behavior from above.
[sqlfluff:rules:L003]
lint_templated_tokens = True

[sqlfluff:rules:L010]  # Keywords
capitalisation_policy = lower

[sqlfluff:rules:L014]  # Unquoted identifiers
capitalisation_policy = lower
unquoted_identifiers_policy = all

[sqlfluff:rules:L016]
# Setting to True allows us to copy/paste long URLs as comments
ignore_comment_lines = True

[sqlfluff:rules:L030]  # Function names
capitalisation_policy = lower

[sqlfluff:rules:L038]
select_clause_trailing_comma = forbid

[sqlfluff:rules:L040]  # Null & Boolean Literals
capitalisation_policy = lower

[sqlfluff:rules:L042]
# By default, allow subqueries in from clauses, but not join clauses.
forbid_subquery_in = join

```
</p>
</details>

**Comment by GClunies** [comment]
@barrywhart This is the UPPER-CASING issue I brought up in #963. Not quite sure if this id a bug or if something is off in my `.sqlfluff` config?

**Comment by barrywhart** [comment]
Thanks!  I'm not familiar with this area of the code, but someone will probably look at this soon.

**Comment by barrywhart** [comment]
@panasenco: Could this behavior be related to your recent awesome work in PR #838? If so, are you potentially interested in investigating this issue?

Also tagging @dmateusp and @alanmcruickshank, who reviewed the PR.

I could possibly look at this next week, but wanted to first check with folks who may be more familiar with this area.

**Comment by panasenco** [comment]
Sure @barrywhart , I'll take a look at this right now.

**Comment by panasenco** [comment]
The first strange thing I'm seeing is that #838 replaced `capitalisation_policy` with `extended_capitalisation_policy` for rule L014, so that config should have raised errors. Continuing to investigate...

**Comment by panasenco** [comment]
I can confirm now that the issue was that the config used `capitalisation_policy` instead of `extended_capitalisation_policy`. That's totally my fault as I should have made outdated configs throw an error.

<details>
<summary>The fixed file produced with `extended_capitalisation_policy`</summary>

```
-- filename: sl_web_spot_pages.sql



with __dbt__cte__sl_web_user_entitlement_base as (
-- filename: sl_web_user_entitlement_base.sql



/*
Rows in this model are:
    - Individual hits of Segment identify() calls on web or native.

This model can be used to map user entitlement to pages or segment events. This
model does NOT summarize the "subscription window(s)" for a given user_id and
subscription. Further modeling and transformations are required to extract
subscription window summaries.
*/

    with final as (

        select distinct  -- Make sure duplicate rows are not included
            user_id,
            subscription_entitlement,
            received_at as entitlement_from,
            coalesce(
                 lead(
                received_at, 1
            ) over(partition by user_id order by received_at),
            '3001-01-01' 
            ) as entitlement_to

        from surfline.identifies

        where subscription_entitlement is not null


    )

    select * from final
),

__dbt__cte__base_web_pages as (
--filename: base_web_pages.sql



    with final as (

        select
            received_at,
            id as page_id,
            spot_id,
            has_cam,
            user_id,
            anonymous_id,
            name,
            -- fields below are in production as of: 2020-01-31 (1st full day of data)
            cam_id,
            is_single_cam,
            is_mobile_view,
            cam_name_primary

        from surfline.pages


    )

    select * from final
),

user_entitlement as (

    select * from __dbt__cte__sl_web_user_entitlement_base

),

has_cam_entitlement as (

    select * from "segment"."greg_clunies"."spots_has_cam_map"

),

pages as (

    select * from __dbt__cte__base_web_pages

),

pages_xf as (

    select
        pages.received_at,
        pages.page_id,
        pages.spot_id,
        pages.user_id,
        pages.anonymous_id,
        pages.cam_id,
        pages.is_single_cam,
        pages.is_mobile_view,
        pages.cam_name_primary,
        user_entitlement.subscription_entitlement

    from pages
    left join user_entitlement
        on pages.user_id = user_entitlement.user_id
            and pages.received_at >= user_entitlement.entitlement_from
            and pages.received_at < user_entitlement.entitlement_to
    where pages.spot_id is not null
        and pages.name = 'Spot Report'

        -- Remove multicam favorites page hits for now due to varying # of cams on page (user selected)
        -- and pages.context_page_url != 'https://www.surfline.com/surf-cams'
        /*
        2019-12-28
        There are only 13 records of the above "&" scenario - which is immaterial.
        This might not be worth the compute time.

        - Julian
        */



        -- Only get new data (not including today's data)
        and pages.received_at > (
            select
                max(received_at)
            from "segment"."greg_clunies"."sl_web_spot_pages"
        )
        and pages.received_at < current_date



),

final as (

    select
        pages_xf.received_at,
        pages_xf.page_id,
        pages_xf.spot_id,
        has_cam_entitlement.has_cam,
        pages_xf.user_id,
        pages_xf.anonymous_id,
        pages_xf.cam_id,
        pages_xf.is_single_cam,
        pages_xf.is_mobile_view,
        pages_xf.cam_name_primary,
        pages_xf.subscription_entitlement

    from pages_xf
    -- inner join because some spot_ids not in MongoDB
    inner join has_cam_entitlement using (spot_id)

)

select * from final
```

</details>

I will take a little time to fix this properly this time and open a new PR with the fix.

**Comment by GClunies** [comment]
Thanks @panasenco for solving this one! I tested with `extended_capitalisation_policy` and it's linting just like you show in your comment above. ✅

**Comment by barrywhart** [comment]
I'm closing this issue. The PR has not been merged, but that PR adds error reporting when the configuration is not set up correctly, and there is a solution without this. (It's still a good change, though!)
